prompting/validators/criteria.py

```diff
@@ -0,0 +1,96 @@
+# The MIT License (MIT)
+# Copyright © 2023 Yuma Rao
+# Copyright © 2023 Opentensor Foundation
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the “Software”), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+import re
+import torch
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from typing import List
+from enum import Enum
+
+
+@dataclass
+class TaskCriterion(ABC):
+    """
+    Abstract base class for defining task-specific evaluation criteria.
+
+    Attributes:
+        text (str): Text of the criterion to be added to the prompt.
+        penalty (float): Penalty value associated with the criterion.
+    Returns:
+        torch.FloatTensor: Tensor containing the penalty values for each response.
+    """
+
+    text: str
+    penalty: float
+
+    @abstractmethod
+    def evaluate(self, completions: List[str]) -> torch.FloatTensor:
+        pass
+
+    @abstractmethod
+    def compose_text(self) -> str:
+        pass
+
+
+class TextLengthUnitEnum(Enum):
+    CHARACTERS = "characters"
+    WORDS = "words"
+    SENTENCES = "sentences"
+    PARAGRAPHS = "paragraphs"
+
+
+@dataclass
+class MatchLengthCriteria(TaskCriterion):
+    text: str = "Your response must have {target_length} {unit}."
+    penalty: float = 0.1
+    target_length: int = 100
+    unit: TextLengthUnitEnum = TextLengthUnitEnum.WORDS
+
+    def _get_completion_length(self, response: str) -> int:
+        unit_to_split_pattern = {
+            TextLengthUnitEnum.CHARACTERS: None,
+            TextLengthUnitEnum.WORDS: r" +",
+            TextLengthUnitEnum.SENTENCES: r"\. +",
```

Comment:
Perhaps you could include a safeguard against punishing abbreviated names by requiring a minimum sentence length here. I believe 2 words is a reasonable heuristic for minimum sentence length

Reply:
not sure if I understood what you are proposing, this is the split pattern verification, where we split the sentences on `.` . Are you proposing to increment the functionality of the SENTENCES verification by adding a constraint that every sentence should have at least two words to be considered a sentence?

Reply:
exactly. `r'(\s\w+)+\.\s'` or something

Reply:
haha this does it, but so much more complicated:
```python
import re

def count_sentences(text):
    # Regex explanation:
    # \b[A-Z]: Match an uppercase letter at the beginning of a word (assumed start of a sentence).
    # (?:[a-zA-Z]\. ){0,2}: Match 0-2 occurrences of an abbreviation-like pattern (single letter followed by a dot and space).
    # (?:[a-zA-Z]+\s+){1,}: Match at least one word followed by one or more whitespace characters.
    # [a-zA-Z]+: Match the last word before the end punctuation.
    # [.?!]: Match sentence-ending punctuation.
    # (?!\w): Negative lookahead to ensure the punctuation is not followed by an alphanumeric character (part of an abbreviation).
    # The pattern ignores common abbreviations by not counting them as sentence terminators.
    pattern = r'\b[A-Z](?:[a-zA-Z]\. ){0,2}(?:[a-zA-Z]+\s+){1,}[a-zA-Z]+[.?!](?!\w)'

    # Find all matches
    sentences = re.findall(pattern, text)

    # Return the count
    return len(sentences)

# Example usage:
text = "Mr. Smith bought a car. It is a Tesla. He seems quite happy with it!"
print(count_sentences(text))  # Output should be 2, as "Mr." is an abbreviation and should not count as the end of a sentence.
```

Reply:
This is great! Will add this to the code in a bit

Reply:
This pattern was not working properly, text example:
```python
text = "Vancouver, a vibrant city located in the southwestern part of Canada in British Columbia, boasts a rich and varied history. The area was originally inhabited by Indigenous peoples, including the Squamish, Musqueam, and Tsleil-Waututh Nations, for thousands of years before European exploration. In 1792, Captain George Vancouver, a British officer, explored the region, and it was his name that the city would eventually adopt. However, it wasn’t until the mid-19th century that the area saw significant European settlement, spurred by the Fraser Canyon Gold Rush in 1858. The settlement grew rapidly, and in 1886, the city was incorporated just days before the arrival of the first transcontinental train, connecting Vancouver to the rest of Canada. This key development helped to establish the city as a major port and economic hub. Over the decades, Vancouver has continued to grow and evolve, welcoming immigrants from all over the world and developing a reputation for its diverse culture, stunning natural scenery, and high quality of life. Today, it stands as Canada’s third-largest metropolitan area and a global city with a unique history that shapes its character and identity."
```

- expected count: 8
- real count: 3

The pattern was updated to be 
```python
pattern = re.findall(r"(?<![A-Z])[\.\?!](?:\s|$)", text)
``` 

We can refine more complex use cases in the next release